### ISSUE TITLE
Question about connection to game

### ISSUE THREAD

**Issue by Lerbytech** [Kalaborative/survivio-plus#332]
Hi all and devs especially.
I am data-engineer and want to make dataset of survivio. To do this, I want to record some games. Recording screencasts and transcoding them is too difficult so I wanted to try direct connection to game.
Mainly, I do request to ```http://surviv.io/api/find_game```. And get json like this:
```
 {"res":[{"zone":"waw","gameId":"53602255248287e3ec89cd00f9e4cda0","hosts":["eu-waw-01.surviv.io:449","eu-waw-01.surviv.io"],"addrs":["54.38.194.11:444","54.38.194.11"],"data":"ksErP4tovbwUtVpO4iSF4guHzJa7CKjHFqfPGENhih0elF8W9qZFHwIeIdvKsErDTg7iMqFurN+McUXW49OyPwi2LBH0xkWZeFyJudGCnKE8WwoFtjKQRmULb22mCnKi9ZfQPNwMWapYCiNxRvDUts1YriBe93Sgm0WV0CzXC5Fn+H4uNzP4WaFjR3JlgqolbdprgrXeyuGgC7tZ5FO4OrJDd/u96p/75G8dUSUCB1tffZ3OgLdXZ7LNpMe3eVpNssG0BDPJGItu","useHttps":false}]}
```
Data from this json is used to connect to survivio server via websocket protocol. 
But I can't find how to make this data to websocket request. For aforementioned json my browser creates this request (coded as curl):
```
curl 'ws://54.38.194.11:444/play?gameId=53602255248287e3ec89cd00f9e4cda0' -H 'Pragma: no-cache' -H 'Origin: http://surviv.io' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7' -H 'Sec-WebSocket-Key: +qLjgICDV4BwAw9UnVJEWQ==' -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.96 Safari/537.36' -H 'Upgrade: websocket' -H 'Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits' -H 'Cache-Control: no-cache' -H 'Connection: Upgrade' -H 'Sec-WebSocket-Version: 13' --compressed
```
How to cast Sec-WebSocketKey from 'data' field in json? Any ideas or advice?
I've read that this Key is established during client-server handshake procedure. 
But when I tried using websocket-client library to make direct connection to server with known game id (that way handshake will be completely new as server should count me as another new player), connection was established and stopped after several packets. I guess in json some additional data is encoded, but don't know how to interpret it. 
I checked whole repository, but as python developer I am not good in js and no direct answer was found.

As another option I guess one can modify cheat source code so all incomin messages would be logged to file. This is also would be also fine to me. Could anyone show me which lines in app should be modified?

I appreciate any help. thanks in advance

**Comment by iBLiSSINYT** [comment]
Ok, I will summon Kalaborative

**Comment by Kalaborative** [comment]
If you mean modifying the game's source code, see `background.js`: you can search the source code and replace with a `console.log` using regex
I'm unfamiliar with how to mock websocket requests so I'm not sure how to help there... my guess is the server is attaching an encoded session key before making the request, but I don't know how those keys are generated
you would need to run the javascript in a Node environment to write logging messages to a file, AFAIK

In short, I don't really know. 😂 
The game's source code is so huge it would be hard to pinpoint what data would be relevant from a data analyzation standpoint.

**Comment by IceHacks** [comment]
@Lerbytech I don't think any of us are data engineers, so.... do you want to connect to the WebSocket? Or just record games? I don't really know which, but I recommend NodeJS and using a headless chrome like `puppeteer` to record the gameplay. Or if you just want to watch the WebSocket use a web-debugger like fiddler or the chrome devtools and that will let you inspect all the WebSocket traffic. 

What do you hope to accomplish by doing this?

**Comment by zbot473** [comment]
@Lerbytech Is this for an ML project? If so, I am really interested. I know how to inspect websocket data, but a github issue comment is really not the medium that i can use.

**Comment by zbot473** [comment]
If you have some IM service or discord, I can message you there.

**Comment by iBLiSSINYT** [comment]
![](https://i.imgflip.com/2vj295.jpg)

**Comment by zbot473** [comment]
Reverse engineering the ws is relatively difficult, so you may want to create an extension that give you the readout of the data. I can help with that
